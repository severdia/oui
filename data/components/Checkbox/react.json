{"description":"Generates a `checkbox` element wrapped in a Label","methods":[],"props":{"checked":{"type":{"name":"bool"},"required":false,"description":"Boolean to set checkbox, for controlled component"},"defaultChecked":{"type":{"name":"bool"},"required":false,"description":"Boolean for how checkbox renders initially"},"isDisabled":{"type":{"name":"bool"},"required":false,"description":"Prevents checkbox from being modified and appears disabled"},"label":{"type":{"name":"string"},"required":false,"description":"Text that describes the checkbox"},"onChange":{"type":{"name":"func"},"required":false,"description":"Function that fires when the checkbox is clicked"},"testSection":{"type":{"name":"string"},"required":false,"description":"Hook for automated JavaScript tests"}},"private":false}