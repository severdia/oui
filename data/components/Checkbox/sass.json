{"name":"Checkbox\n","description":"","start":{"line":1,"column":1},"end":{"line":14,"column":4},"example":[{"type":"html","description":null,"code":"<ul class=\"oui-input-list\">\n  <li>\n    <input name=\"checkbox-set-1\" id=\"bar-1\" type=\"checkbox\" checked>\n    <label class=\"oui-label\" for=\"bar-1\">Checkbox Option #1\n  </li>\n  <li>\n    <input name=\"checkbox-set-1\" id=\"bar-2\" type=\"checkbox\">\n    <label class=\"oui-label\" for=\"bar-2\">Checkbox Option #2\n  </li>\n</ul>"}]}