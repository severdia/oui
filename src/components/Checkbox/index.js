import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import Label from '../Label';

/**
 * Generates a `checkbox` element wrapped in a Label
 * @param {Object} props - Properties passed to component
 * @returns {ReactElement}
 */
const Checkbox = ({
  checked,
  defaultChecked,
  isDisabled,
  label,
  labelWeight,
  onChange,
  testSection,
}) => {
  const labelClassNames = classNames({
    'flush--bottom': true,
    'push--left': true,
    'weight--light': labelWeight === 'light',
    'weight--normal': labelWeight === 'normal',
    'weight--bold': labelWeight === 'bold',
    'cursor--pointer': true,
    'oui-label--disabled': isDisabled,
  });
  const classes = classNames({
    'flex--none': true,
  });
  return (
    <Label testSection={ testSection && testSection + '-label' }>
      <div className="flex">
        <input
          data-oui-component={ true }
          type="checkbox"
          defaultChecked={ defaultChecked }
          checked={ checked }
          className={ classes }
          disabled={ isDisabled }
          onChange={ onChange }
          data-test-section={ testSection }
          style={{ 'marginTop': '0.35em' }}
        />
        <div className={ labelClassNames }>
          { label }
        </div>
      </div>
    </Label>
  );
};

Checkbox.propTypes = {
  /** Boolean to set checkbox, for controlled component  */
  checked: PropTypes.bool,
  /** Boolean for how checkbox renders initially  */
  defaultChecked: PropTypes.bool,
  /** Prevents checkbox from being modified and appears disabled */
  isDisabled: PropTypes.bool,
  /** Label that describes the checkbox */
  label: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
  ]),
  /** Font weight for the label text */
  labelWeight: PropTypes.oneOf(['light', 'normal', 'bold']),
  /** Function that fires when the checkbox is clicked */
  onChange: PropTypes.func,
  /** Hook for automated JavaScript tests */
  testSection: PropTypes.string,
};

Checkbox.defaultProps = {
  labelWeight: 'normal',
};

export default Checkbox;
