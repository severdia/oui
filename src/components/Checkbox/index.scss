////
/// Checkbox
/// @example[html]
///   <ul class="oui-input-list">
///     <li>
///       <input name="checkbox-set-1" id="bar-1" type="checkbox" checked>
///       <label class="oui-label" for="bar-1">Checkbox Option #1
///     </li>
///     <li>
///       <input name="checkbox-set-1" id="bar-2" type="checkbox">
///       <label class="oui-label" for="bar-2">Checkbox Option #2
///     </li>
///   </ul>
////


// This file is only needed to display the examples.
