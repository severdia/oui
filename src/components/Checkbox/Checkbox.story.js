import React from 'react';

import { storiesOf } from '@storybook/react';
import { withKnobs, text, select, boolean } from '@storybook/addon-knobs';
import Checkbox from '../Checkbox';

const stories = storiesOf('Checkbox', module);
stories
  .addDecorator(withKnobs)
  .addDecorator(story => (
    <div id="root-preview">
      {story()}
    </div>
  ));

stories
  .add('With string label', (() => {
    return (
      <Checkbox
        defaultChecked={ true }
        label={ text('label', 'checkbox component') }
        isDisabled={ boolean('isDisabled', false) }
      />
    );
  }))
  .add('With html label', (() => {
    return (
      <Checkbox
        defaultChecked={ true }
        label={
          <div>
            <strong>Label</strong>
            <br />
            <span>Description text</span>
          </div>
        }
        isDisabled={ boolean('isDisabled', false) }
      />
    );
  }))
  .add('With labelWeight', (() => {
    return (
      <Checkbox
        defaultChecked={ true }
        label={ text('label', 'checkbox component') }
        labelWeight={ select('labelWeight', ['light', 'normal', 'bold'], 'bold') }
        isDisabled={ boolean('isDisabled', false) }
      />
    );
  }));
