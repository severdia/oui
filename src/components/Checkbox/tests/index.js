import React from 'react';
import Checkbox from '../index';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

describe('components/Checkbox', () => {
  const mockFunction = jest.fn();
  it('should render an input of type checkbox', () => {
    const component = mount(<Checkbox />);
    expect(component.find('input[type="checkbox"]').length).toBe(1);
  });

  it('should render a checked checkbox when `defaultChecked` is true', () => {
    const component = shallow(<Checkbox defaultChecked={ true } />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should render an unchecked checkbox when `defaultChecked` is false', () => {
    const component = shallow(<Checkbox defaultChecked={ false } />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should render a checked checkbox when `checked` is true', () => {
    const component = shallow(<Checkbox checked={ true } onChange={ mockFunction } />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should render an unchecked checkbox when `checked` is false', () => {
    const component = shallow(<Checkbox checked={ false } onChange={ mockFunction } />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should throw an error if a checkbox is rendered without an onChange function', () => {
    spyOn(console, 'error').and.stub();

    mount(<Checkbox checked={ true } />);

    expect(console.error.calls.all()[0].args[0]).toContain('You provided a `checked` prop to a form field without an `onChange` handler'); // eslint-disable-line
  });

  it('should call the onChange event handler when the checkbox is changed', () => {
    const handler = {
      onChange: (event) => {},
    };
    spyOn(handler, 'onChange');

    const component = shallow(<Checkbox onChange={ handler.onChange } />);

    component.find('input').simulate('change');
    expect(handler.onChange).toHaveBeenCalled();
  });

  it('should be wrapped in a label', () => {
    const component = mount(<Checkbox />);

    const label = component.find('label');
    const checkbox = label.find('input[type="checkbox"]');

    expect(label.length).toBe(1);
    expect(checkbox.length).toBe(1);
  });

  it('label should render text', () => {
    const component = mount(<Checkbox label="foo" />);

    const label = component.find('label');
    expect(label.text()).toBe('foo');
  });

  it('should add disabled class to label text when disabled', () => {
    const component = mount(<Checkbox isDisabled={ true } />);

    const label = component.find('.oui-label--disabled');
    expect(label.length).toBe(1);
  });

  it('should not add disabled class to label text when not disabled', () => {
    const component = mount(<Checkbox isDisabled={ false } />);

    const label = component.find('.oui-label--disabled');
    expect(label.length).toBe(0);
  });

  it('should add disabled attribute to checkbox when disabled', () => {
    const component = mount(<Checkbox isDisabled={ true } />);

    const input = component.find('input[disabled]');
    expect(input.length).toBe(1);
  });

  it('should add weight--normal class to label text when no labelWeight prop present', () => {
    const component = mount(<Checkbox />);

    const label = component.find('.weight--normal');
    expect(label.length).toBe(1);
  });

  it('should add weight--normal class to label text when labelWeight prop = "normal"', () => {
    const component = mount(<Checkbox labelWeight="normal" />);

    const label = component.find('.weight--normal');
    expect(label.length).toBe(1);
  });

  it('should add weight--light class to label text when labelWeight prop = "light"', () => {
    const component = mount(<Checkbox labelWeight="light" />);

    const label = component.find('.weight--light');
    expect(label.length).toBe(1);
  });

  it('should add weight--bold class to label text when labelWeight prop = "bold"', () => {
    const component = mount(<Checkbox labelWeight="bold" />);

    const label = component.find('.weight--bold');
    expect(label.length).toBe(1);
  });

  it('should not add disabled attribute to checkbox when not disabled', () => {
    const component = mount(<Checkbox isDisabled={ false } />);
    expect(component.find('input').prop('disabled')).toBe(false);
  });

  it('should properly add a test section to label', () => {
    const component = mount(<Checkbox testSection="duck" />);
    expect(component.find('[data-test-section="duck-label"]').length).toBe(1);
  });

  it('should properly add a test section to checkbox', () => {
    const component = mount(<Checkbox testSection="duck" />);
    expect(component.find('input[type="checkbox"][data-test-section="duck"]').length).toBe(1);
  });
});
