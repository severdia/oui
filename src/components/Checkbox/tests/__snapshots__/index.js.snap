// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Checkbox should render a checked checkbox when \`checked\` is true 1`] = `
<Label>
  <div
    className="flex"
  >
    <input
      checked={true}
      className="flex--none"
      data-oui-component={true}
      onChange={[MockFunction]}
      style={
        Object {
          "marginTop": "0.35em",
        }
      }
      type="checkbox"
    />
    <div
      className="flush--bottom push--left weight--normal cursor--pointer"
    />
  </div>
</Label>
`;

exports[`components/Checkbox should render a checked checkbox when \`defaultChecked\` is true 1`] = `
<Label>
  <div
    className="flex"
  >
    <input
      className="flex--none"
      data-oui-component={true}
      defaultChecked={true}
      style={
        Object {
          "marginTop": "0.35em",
        }
      }
      type="checkbox"
    />
    <div
      className="flush--bottom push--left weight--normal cursor--pointer"
    />
  </div>
</Label>
`;

exports[`components/Checkbox should render an unchecked checkbox when \`checked\` is false 1`] = `
<Label>
  <div
    className="flex"
  >
    <input
      checked={false}
      className="flex--none"
      data-oui-component={true}
      onChange={[MockFunction]}
      style={
        Object {
          "marginTop": "0.35em",
        }
      }
      type="checkbox"
    />
    <div
      className="flush--bottom push--left weight--normal cursor--pointer"
    />
  </div>
</Label>
`;

exports[`components/Checkbox should render an unchecked checkbox when \`defaultChecked\` is false 1`] = `
<Label>
  <div
    className="flex"
  >
    <input
      className="flex--none"
      data-oui-component={true}
      defaultChecked={false}
      style={
        Object {
          "marginTop": "0.35em",
        }
      }
      type="checkbox"
    />
    <div
      className="flush--bottom push--left weight--normal cursor--pointer"
    />
  </div>
</Label>
`;
